nixos/ohMyZsh: allow multiple derivations in `ZSH_CUSTOM`

Ma27 opened the pull request:
###### Motivation for this change

For third-party scripts `oh-my-zsh` provides `ZSH_THEME`. However this can be set to only one directory. However the module should be able to support multiple third-party scripts packaged as derivations using nix's composability features (namely `buildEnv`).

Now a config like this is possible:

``` nix
{ pkgs, ... }:
{
  programs.zsh.enable = true;
  programs.zsh.ohMyZsh = {
    enable = true;
    customPkgs = [
      pkgs.lambda-mod-zsh-theme
      pkgs.nix-zsh-completions
    ];
    theme = "lambda-mod";
    plugins = [ "nix" ];
  };
}
```

The support for `ohMyZsh.custom` has been retained as it's possible that `ZSH_CUSTOM` should point to a personal directory (e.g. a scripts dir in `$HOME`). An assertion ensures that no collisions between these features happens.

The commits are structured like this:

* modify `nix-zsh-completions` package (add hooks, cleanup meta, add plugin script, allow (optionally) an oh-my-zsh like structure (with `themes/`,`plugins/` etc rather than `share/zsh`)
* improve `lambda-mod-zsh-theme` (only usable with `oh-my-zsh`, drops usage of `$out/share`, instead `$out/themes` is used)
* relevant updates to the modules

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Ma27 committed:
nix-zsh-completions: install nix.plugin.zsh and init.zsh as well

These files feature simple aliases and a nix-shell detector when using
`nix-shell --run zsh`. The package itself contains the completion scripts in
`$out/share/zsh/site-functions` (to keep it compatible with ZSH-only
setups) and the plugins in `$out/share/zsh/plugins` for oh-my-zsh.

Ma27 committed:
lambda-mod-zsh-theme: add hooks and patch shebangs

This ensures that no impurity exists due to implicitly depending
on`/usr/bin/env`.

It stores the theme into `$out/share/zsh/themes` to make it possible to
find theme using `buildEnv` and remain consistent with other ZSH
extensions.

Ma27 committed:
nixos/oh-my-zsh: add `customPkgs` option to allow multiple derivations for `ZSH_CUSTOM`

If multiple third-party modules shall be used for `oh-my-zsh` it has to
be possible to create another env which composes all the packages.

Now it can be done like this:

```
{ pkgs, ... }:
{
  programs.zsh.enable = true;
  programs.zsh.ohMyZsh = {
    enable = true;
    customPkgs = with pkgs; [
      lambda-mod-zsh-theme
      nix-zsh-completions
    ];
    theme = "lambda-mod";
    plugins = [ "nix" ];
  };
}
```

Please keep in mind that this is not compatible with
`programs.zsh.ohMyZsh.custom`, only one of these options can be used
ATM.

Each package should store its outputs into
`$out/share/zsh/<output-name>`. Completions (and ZSH-only) extensions
should live in the `fpath` (`$out/share/zsh/site-functions`), plugins in
`.../plugins` and themes in `.../themes` (please refer to
fdb6bf6ed68c2f089ae6c729dfeaa3eddea2ce6a and 406d64aad162b3a4881747be4e24705fb5182573).

All scripts in `customPkgs` will be linked together using `linkFarm` to
provide a single directory for all scripts from all derivations in
`customPkgs` as suggested in https://github.com/NixOS/nixpkgs/pull/43282#issuecomment-410396365.

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=35fb99fe-d281-43d1-8bc5-bdd87968d788)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/7i5ym1srqq454abdc0k1ahs4vpn4j708-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/px9fz6lsqs5v9q3kmrmr1ckgmawxkdsm-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/4qvrxzxa535y8304mk195x50b6p9607d-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/px9fz6lsqs5v9q3kmrmr1ckgmawxkdsm-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/px9fz6lsqs5v9q3kmrmr1ckgmawxkdsm-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/px9fz6lsqs5v9q3kmrmr1ckgmawxkdsm-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/blrbh6scgmchsafg6gv8mmaabnj8k56f-nix-zsh-completions-0.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=bd5d6673-38ec-4e7e-b039-8c9ea35432f3)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/l322yvzki4qlkjlfqfinlxsqj0j6rpy5-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/9x2gwqwr47qv0bazh6dmwg1vxbj8ivl9-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/0pjsgkxz0rp5baycq5sp2s72lrr5q9sg-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/9x2gwqwr47qv0bazh6dmwg1vxbj8ivl9-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/9x2gwqwr47qv0bazh6dmwg1vxbj8ivl9-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/9x2gwqwr47qv0bazh6dmwg1vxbj8ivl9-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/kb9dzgvh70k3jfmvrnqmwzs5xkyzvmsf-nix-zsh-completions-0.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=3c37d101-cabb-440e-a551-9d57f058cc48)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/7i5ym1srqq454abdc0k1ahs4vpn4j708-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/dq10a0cmxwbinvml98ff38ryahcpyccs-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/4qvrxzxa535y8304mk195x50b6p9607d-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/dq10a0cmxwbinvml98ff38ryahcpyccs-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/dq10a0cmxwbinvml98ff38ryahcpyccs-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/dq10a0cmxwbinvml98ff38ryahcpyccs-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/blrbh6scgmchsafg6gv8mmaabnj8k56f-nix-zsh-completions-0.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=7155fef5-f21e-48bc-b4ce-923628cf659c)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/l322yvzki4qlkjlfqfinlxsqj0j6rpy5-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/lf2lyaq00lhfrlxyjgg8igxkl41lk8mg-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/0pjsgkxz0rp5baycq5sp2s72lrr5q9sg-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/lf2lyaq00lhfrlxyjgg8igxkl41lk8mg-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/lf2lyaq00lhfrlxyjgg8igxkl41lk8mg-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/lf2lyaq00lhfrlxyjgg8igxkl41lk8mg-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/kb9dzgvh70k3jfmvrnqmwzs5xkyzvmsf-nix-zsh-completions-0.4.0
```
</p></details>

Infinisil commented on pkgs/shells/zsh/nix-zsh-completions/default.nix:
```diff
@@ -14,16 +14,26 @@ stdenv.mkDerivation rec {
-    cp _* $out/share/zsh/site-functions
+  installPhase = let
+    pluginDir = if ohMyZsh then "plugins/nix" else "share/zsh/site-functions";
+    compDir   = if ohMyZsh then "completions" else "share/zsh/site-functions";
```
Is this way of packaging ohMyZsh plugins documented somewhere? I can't find anything else in nixpkgs using this, and it feels like that shouldn't be part of the package declaration, but rather of the code that wants to use the package (by wrapping the paths how ohmyzsh expects them or so).

Infinisil commented on nixos/modules/programs/zsh/oh-my-zsh.nix:
```diff
@@ -34,10 +34,18 @@ in
+          default = null;
+          type = with types; nullOr str;
+          description = ''
+           List of custom packages that should be loaded into `oh-my-zsh`.
```
The description is the same here as the new option below.

Infinisil commented on nixos/modules/programs/zsh/oh-my-zsh.nix:
```diff
@@ -60,23 +68,32 @@ in
     };
 
-    config = mkIf cfg.enable {
+    config = with builtins; mkIf cfg.enable (let
```
`with builtins;` can be removed, `length` and `head` are in `lib`.

Infinisil commented on nixos/modules/programs/zsh/oh-my-zsh.nix:
```diff
@@ -60,23 +68,32 @@ in
+
+      custom = if cfg.custom != null then cfg.custom else
+        if length (cfg.customPkgs) == 0 then null else
+        if length (cfg.customPkgs) == 1 then head cfg.customPkgs else pkgs.buildEnv {
```
Nit: Parens can be removed

Ma27 commented:
@Infinisil thanks a lot for your review! I fixed all of it and solved theproblem with oh-my-zsh packages by adjusting an attr set in the `passthru` value of the corresponding derivation which maps out dirs to oh-my-zsh dirs to build proper packages, but provide information on how to structure the output in an oh-my-zsh environment.

Infinisil commented:
Thanks, but I still don't think that's quiet right. (Asking because I'm not entirely sure:) So oh-my-zsh expects `ZSH_CUSTOM` to be set to a folder, containing subfolders for `themes` and `completions`. In `completions` there will be files like `_foo` for a completion for command `foo`, this is entirely usable with zsh and doesn't require oh-my-zsh and seems to be entirely what the `$out/share/zsh/site-functions` folder is dedicated for. In `themes` there seem to be themes also usable entirely by zsh itself, no oh-my-zsh needed, but there doesn't seem to be a dedicated folder for this. Is this correct?

If that's the case, then I suggest to dedicate a new folder `$out/share/zsh/themes` for zsh themes and make theme packages install stuff there (and potentially make it backwards compatible). Creating the zsh custom package can then look like this:

```nix
linkFarm "oh-my-zsh-custom" [
  {
    name = "themes";
    path = "${buildEnv {
      name = "oh-my-zsh-themes";
      paths = [ customPkgs ];
      pathsToLink = "/share/zsh/themes";
    }}/share/zsh/themes";
  }
  {
    name = "completions";
    path = "${buildEnv {
      name = "oh-my-zsh-completions";
      paths = [ customPkgs ];
      pathsToLink = "/share/zsh/site-functions";
    }}/share/zsh/site-functions";
  }
]
```

And a thing about the current state: `{pre,post}Install` hooks are usually not ran manually if not needed. I'm not exactly sure why, but almost aren't any derivations doing this in nixpkgs.

Ma27 commented:
Thanks for the review and the detailed feedback! I mostly agree with you, but now I have to catch some sleep, will have  a deeper look at this tomorrow %)

Ma27 commented:
@Infinisil fixed, thanks a lot for your feedback!

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=c6f027c0-9d33-4f76-9ad7-ca74ed322c64)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/9lfz6xj30b45xmyingcix77f3s1gp3ay-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/i9l5m6g8914hh3jlamzqglyizx5dmh99-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/1hi76hr87bd1y1q1qjk0lv8nmcjip1c8-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/i9l5m6g8914hh3jlamzqglyizx5dmh99-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/i9l5m6g8914hh3jlamzqglyizx5dmh99-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/i9l5m6g8914hh3jlamzqglyizx5dmh99-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/1jn5b5zysxxg8yx1b45605mq6r1gnsic-nix-zsh-completions-0.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=31564918-e294-407b-b75c-200116c5182f)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/86bybqzriqfihqic4bxghrj8ww2ljk1d-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/if5vldckglf5rvdl9b7pf2daiacjsikf-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/zrs21zqcchgyabjf4xfimncdq16njizc-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/if5vldckglf5rvdl9b7pf2daiacjsikf-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/if5vldckglf5rvdl9b7pf2daiacjsikf-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/if5vldckglf5rvdl9b7pf2daiacjsikf-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/nsq86kxmbgr5wwg6qfxv074py8l3y220-nix-zsh-completions-0.4.0
```
</p></details>

Infinisil commented on pkgs/shells/zsh/nix-zsh-completions/default.nix:
```diff
@@ -15,15 +15,20 @@ stdenv.mkDerivation rec {
     cp _* $out/share/zsh/site-functions
+    cp *.zsh $out/share/zsh/plugins/nix
+
+    runHook postInstall
```
As previously mentioned, don't run these hooks manually here (for both derivations).

Infinisil commented on pkgs/shells/zsh/lambda-mod-zsh-theme/default.nix:
```diff
@@ -10,9 +10,18 @@ stdenv.mkDerivation {
+
+    chmod +x lambda-mod.zsh-theme # only executable scripts are found by `patchShebangs`
+    patchShebangs .
+    chmod -x lambda-mod.zsh-theme
```
I think it's okay to let it stay executable, it is executable after all, especially since the shebang was patched.

infinisil reviewed (changes requested):
Looking much better! With this PR, these `/share/zsh/{themes,site-functions,plugins/*}` directories will probably become a convention for how to package zsh stuff on Nix, so I think it deserves an entry in the docs. Would you mind adding such an entry? Would be somewhere in the Nixpkgs manual. This would then be something we can link people to if they want to package zsh stuff and have it work with nixpkgs' conventions.

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=ee69ff1a-8918-40bb-8244-74a10207e301)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/igq8lbcfcxmmkcsrcv39wr39n7wrr1pm-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/l2fbkaa0cgmfs4qx80hvllrlw29fy4mi-nix-zsh-completions-0.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=8b500d88-554b-4c57-b0b7-6c04d381f1a0)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
/nix/store/c64wyvxa22pynl3mzp6waq28ndjm8wzx-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/g5s02zis189wm8c1iyv8vgw8kj3lqlfk-nix-zsh-completions-0.4.0
```
</p></details>

Ma27 commented:
cc @Infinisil, are the latest changes better?

Infinisil commented:
Yeah looking good now! Only thing that would be nice is some docs, documenting `$out/share/zsh/{themes,site-functions,plugins}`: What each of them is for, what the directories should look like. Currently only `$out/share/zsh/site-functions` is a "known" (but not documented?) convention, probably due to `make install`s populating that dir.

Adding docs should prevent anybody else trying to define some directory as a convention for themes/plugins. I think this may be the reason we currently need to support multiple of those for completions: https://github.com/NixOS/nixpkgs/blob/a2f499e21df9aa447e6a7af2be030a58447f877f/nixos/modules/programs/zsh/zsh.nix#L157-L160 (or it might be because there already were multiple conventions for the make installs)

Ma27 commented:
great!

You're absolutely right, it could use some documentation, please give me some time to think how and where to write it, I'll report back, when it's done.

And thanks for your patience and feedback! :)

Ma27 committed:
nixos/oh-my-zsh: add documentation

In the last year `programs.oh-my-zsh` gained more complexity and since
the introduction of features like `customPkgs` which builds a
`ZSH_CUSTOM` path from a sequence of derivation a documentation may be
fairly helpful to make the knowledge how to use the module and how to
package new ZSH plugins visible.

See https://github.com/NixOS/nixpkgs/pull/43282#issuecomment-410770432

Ma27 commented:
@Infinisil I added a docbook entry (using `meta.doc`) for `oh-my-zsh`. I thought it might be worth documenting the entire module and added one paragraph which describes how to package third-party oh-my-zsh extensions.

GrahamcOfBorg commented:
Success on aarch64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=350301bb-4423-480f-8243-5f304815a225)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/86bybqzriqfihqic4bxghrj8ww2ljk1d-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/c64wyvxa22pynl3mzp6waq28ndjm8wzx-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/zrs21zqcchgyabjf4xfimncdq16njizc-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/c64wyvxa22pynl3mzp6waq28ndjm8wzx-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/c64wyvxa22pynl3mzp6waq28ndjm8wzx-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/c64wyvxa22pynl3mzp6waq28ndjm8wzx-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/g5s02zis189wm8c1iyv8vgw8kj3lqlfk-nix-zsh-completions-0.4.0
```
</p></details>

GrahamcOfBorg commented:
Success on x86_64-linux [(full log)](https://logs.nix.ci/?key=nixos/nixpkgs.43282&attempt_id=75256463-f463-42bf-83b4-e9e02ecaa86c)

Attempted: lambda-mod-zsh-theme, nix-zsh-completions

<details><summary>Partial log (click to expand)</summary><p>

```
installing
patching script interpreter paths in .
./lambda-mod.zsh-theme: interpreter directive changed from "/usr/bin/env zsh" to "/nix/store/9lfz6xj30b45xmyingcix77f3s1gp3ay-zsh-5.5.1/bin/zsh"
post-installation fixup
shrinking RPATHs of ELF executables and libraries in /nix/store/igq8lbcfcxmmkcsrcv39wr39n7wrr1pm-lambda-mod-zsh-theme-unstable-2017-10-08
strip is /nix/store/1hi76hr87bd1y1q1qjk0lv8nmcjip1c8-binutils-2.30/bin/strip
patching script interpreter paths in /nix/store/igq8lbcfcxmmkcsrcv39wr39n7wrr1pm-lambda-mod-zsh-theme-unstable-2017-10-08
checking for references to /build in /nix/store/igq8lbcfcxmmkcsrcv39wr39n7wrr1pm-lambda-mod-zsh-theme-unstable-2017-10-08...
/nix/store/igq8lbcfcxmmkcsrcv39wr39n7wrr1pm-lambda-mod-zsh-theme-unstable-2017-10-08
/nix/store/l2fbkaa0cgmfs4qx80hvllrlw29fy4mi-nix-zsh-completions-0.4.0
```
</p></details>

Infinisil commented:
Neat! 2 things about the docs: It would be nice for them to be built into the nixpkgs docs, so it should probably be in `nixpkgs/docs/manual`. And as far as I know, all of the things in this document apply more generally to zsh, there's nothing oh-my-zsh specific, so I think this should be changed to reflect that.

Thanks for doing all this work :)

Ma27 commented:
> It would be nice for them to be built into the nixpkgs docs

Care to elaborate? I just followed the convention to use `meta.doc` for module-specific documentation which will be rendered into the NixOS manual and I don't understand ATM why it's preferable to put it into the `nixpkgs` docs.

> And as far as I know, all of the things in this document apply more generally to zsh, there's nothing oh-my-zsh specific, so I think this should be changed to reflect that.

Well, plugins are mostly oh-my-zsh specific and many (third-party) themes explicirtly depend on oh-my-zsh. I mentioned that packages should be kept compatible with zsh-only setups (if possible), but I'll re-read the paragraphs and think where we can make this more explicit.

However I'll wait with this until we have decided whether or not to move the docs to `nixpkgs` to finish all of this in a single iteration :)

Infinisil commented:
Oh I see, yeah I didn't know about `meta.doc` actually, sorry! Then I don't have anything to say anymore, this looks good to me :)

Ma27 commented:
@Infinisil any chance to get this merged?:)
